File: apps/ios/mynewsapp/ViewModels/ArticlesViewModel.swift
import Foundation
import Observation

enum DataState {
    case idle
    case loading
    case loaded([Article])
    case error(Error)
}

@Observable
@MainActor
final class ArticlesViewModel {

    var dataState: DataState = .idle
    
    private let fetchAction: () async throws -> [Article]
    
    init(fetchAction: @escaping () async throws -> [Article]) {
        self.fetchAction = fetchAction
    }
    
    func loadData() async {
        dataState = .loading
        AppLogger.viewModels.info("Loading articles from server...")
        do {
            let articles = try await fetchAction()
            dataState = .loaded(articles)
            AppLogger.viewModels.info("Successfully fetched \(articles.count) articles")
        } catch {
            dataState = .error(error)
            AppLogger.viewModels.error("Failed to fetch articles. localizedDescription: \(error.localizedDescription)")
        }
    }
}

File: apps/android/app/src/main/java/com/pouriaalmassi/mynewsapp/viewmodels/ArticlesViewModel.kt
package com.pouriaalmassi.mynewsapp.viewmodels

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.pouriaalmassi.mynewsapp.models.Article
import kotlinx.coroutines.launch

sealed class DataState {
    object Idle : DataState()
    object Loading : DataState()
    data class Loaded(val articles: List<Article>) : DataState()
    data class Error(val error: Throwable) : DataState()
}

class ArticlesViewModel(
    private val fetchAction: suspend () -> List<Article>
) : ViewModel() {

    var dataState by mutableStateOf<DataState>(DataState.Idle)
        private set

    fun loadData() {
        viewModelScope.launch {
            dataState = DataState.Loading
            try {
                val articles = fetchAction()
                dataState = DataState.Loaded(articles)
            } catch (e: Exception) {
                dataState = DataState.Error(e)
            }
        }
    }
}
